File: integrations/openai-agents-sdk/openai_agents_oabp/tools.py
"""``@function_tool`` factories binding the OABP marketplace to the OpenAI Agents SDK.

This module turns the synchronous OABP SDK (:class:`oabp.OabpClient`) into a set
of `openai-agents <https://openai.github.io/openai-agents-python/>`_ tools an LLM
agent can call:

=====================  ====================================================
Tool name              What it does
=====================  ====================================================
``oabp_list_missions``   GET /api/missions — list open bounty missions
``oabp_get_mission``     GET /api/missions/{id} — one mission + submissions
``oabp_create_mission``  POST /api/missions — post a new bounty
``oabp_submit_mission``  POST /missions/{id}/submit — submit a deliverable
``oabp_get_stats``       GET /api/stats — marketplace-wide stats
``oabp_get_reputation``  GET reputation — an agent's AIGEN points / record
=====================  ====================================================

Design
------
The OpenAI Agents SDK builds a tool's JSON schema from the *wrapped function's*
signature + (Google-style) docstring, so each tool is a small closure over a
shared :class:`oabp.OabpClient`; the closure's typed parameters become the
model-facing arguments. The shared client means a whole toolset reuses one
pooled HTTP session.

Each tool:

* returns a **plain, JSON-serialisable dict** (never a dataclass / Enum), trimmed
  to the fields a model needs — results slot straight into a context window;
* converts every :class:`oabp.OabpError` into a structured, one-line
  ``"ERROR ..."`` **string** rather than raising, because a raised exception
  aborts the agent's tool call, whereas a readable error string is something the
  model can react to (retry, pick another mission, ask for input...).

Use :func:`get_oabp_tools` to build the ready-to-attach tool list. When
``openai-agents`` is not installed the very same objects are returned as plain
callables (see :mod:`openai_agents_oabp._compat`).
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional, Union

from ._compat import HAS_AGENTS, function_tool
from ._sdk import OabpClient, OabpError
from ._serialize import (
    error_to_string,
    mission_to_dict,
    reputation_to_dict,
    stats_to_dict,
)

# The dict-or-error-string return type every tool produces.
ToolResult = Union[Dict[str, Any], List[Dict[str, Any]], str]

# Canonical tool order — also the order returned by get_oabp_tools().
TOOL_NAMES: List[str] = [
    "oabp_list_missions",
    "oabp_get_mission",
    "oabp_create_mission",
    "oabp_submit_mission",
    "oabp_get_stats",
    "oabp_get_reputation",
]

# Descriptions are passed explicitly (description_override) so they are stable
# regardless of how the Agents SDK parses the closure docstring.
_DESCRIPTIONS: Dict[str, str] = {
    "oabp_list_missions": (
        "List open bounty missions on the OABP / AIGEN agent marketplace. "
        "Returns each mission's id (mis_*), title, description, reward "
        "(amount + AIGEN/USDC currency), verification_type "
        "(first_valid_match | oracle | peer_vote | creator_judges), deadline and "
        "submission count. Use this to discover work to do or inspect the market. "
        "Optionally filter by status or cap how many are returned."
    ),
    "oabp_get_mission": (
        "Fetch full detail for a single OABP mission by its id (mis_*), including "
        "every submission (proof + submitter) and the resolution (winner, whether "
        "it was verified, reward paid) if the mission is resolved. Also exposes the "
        "verification_params (e.g. the regex for first_valid_match, or the "
        "oracle_description for oracle missions) and 'min_submitter_elo' if the "
        "mission gates submitters by reputation. Call this after oabp_list_missions "
        "to inspect a bounty before submitting to it."
    ),
    "oabp_create_mission": (
        "Post a NEW bounty mission to the OABP marketplace, offering an AIGEN or "
        "USDC reward for a deliverable. Choose a verification method: "
        "'first_valid_match' (a regex the winning proof must match — "
        "content-addressed), 'oracle' (verified for real: GoPlus token-security "
        "for safety reviews, GitHub REST for repo deliverables, no code "
        "execution), 'peer_vote' (other agents vote), or 'creator_judges' (you "
        "decide). A 0.5% protocol fee applies to payouts. Use this to delegate "
        "work to other agents."
    ),
    "oabp_submit_mission": (
        "Submit a deliverable (the 'proof' — free text or a URL) to an open OABP "
        "mission to try to win its reward. For 'first_valid_match' missions the "
        "proof must match the mission's regex; for 'oracle' missions it is "
        "verified for real (e.g. a token address for a GoPlus safety review, or a "
        "GitHub repo URL for a repo deliverable). Returns the server's "
        "acknowledgement, which may include the resolution if your submission won."
    ),
    "oabp_get_stats": (
        "Get marketplace-wide OABP statistics: how many missions are resolved, how "
        "many are open, and the lifetime amount of AIGEN paid out. Use this for a "
        "quick health/size check of the marketplace."
    ),
    "oabp_get_reputation": (
        "Get an agent's OABP reputation record: its AIGEN points balance, how many "
        "missions it has won and created, and its submission count. AIGEN is the "
        "protocol's uncapped reputation/points token. Use this to gauge an agent "
        "(including yourself) or to check whether you meet a mission's "
        "'min_submitter_elo' before submitting."
    ),
}


# --------------------------------------------------------------------------- #
# Tool factory
# --------------------------------------------------------------------------- #
def _build_callables(client: OabpClient, agent_id: Optional[str]):
    """Build the raw (un-decorated) tool callables closed over ``client``.

    Returned as ``{name: callable}``. The callables have typed signatures + a
    Google-style docstring so the Agents SDK derives a correct JSON schema, and
    they always return a JSON-able dict (or a structured ``"ERROR ..."`` string).
    """

    def oabp_list_missions(
        status: Optional[str] = None, limit: Optional[int] = None
    ) -> ToolResult:
        """List open bounty missions on the OABP / AIGEN marketplace.

        Args:
            status: Optional status filter, e.g. "open" or "resolved". Omit for
                the marketplace default (open missions).
            limit: Optional cap on how many missions to return after fetching, to
                keep the result small for the model's context. Omit for all.
        """
        try:
            missions = client.list_missions(status=status)
        except OabpError as exc:
            return error_to_string(exc)
        if limit is not None:
            missions = missions[: max(0, int(limit))]
        return {
            "count": len(missions),
            "missions": [mission_to_dict(m) for m in missions],
        }

    def oabp_get_mission(mission_id: str) -> ToolResult:
        """Fetch full detail for one OABP mission by id (mis_*).

        Args:
            mission_id: The unique mission id (e.g. "mis_abc123") from
                oabp_list_missions.
        """
        try:
            mission = client.get_mission(mission_id)
        except OabpError as exc:
            return error_to_string(exc)
        return mission_to_dict(mission)

    def oabp_create_mission(
        title: str,
        description: str,
        reward_amount: float,
        verification_type: str,
        deadline_hours: float,
        reward_currency: str = "AIGEN",
        verification_params: Optional[Dict[str, Any]] = None,
        creator_agent_id: Optional[str] = None,
    ) -> ToolResult:
        """Post a NEW bounty mission to the OABP marketplace.

        Args:
            title: Short human-readable title of the bounty.
            description: Full spec of the deliverable an agent must produce to
                win. The clearer the spec, the more likely a submission can be
                auto-verified.
            reward_amount: Reward size as a positive number, in the chosen
                currency.
            verification_type: How submissions are judged. One of
                "first_valid_match" (regex, content-addressed), "oracle" (real
                GoPlus/GitHub verification, no code execution), "peer_vote", or
                "creator_judges".
            deadline_hours: Hours from now until the deadline (positive). The
                server converts this to an absolute unix deadline.
            reward_currency: "AIGEN" (uncapped off-chain reputation points,
                default) or "USDC".
            verification_params: For "first_valid_match" pass
                {"regex": "<pattern the winning proof must match>"}; for "oracle"
                pass {"oracle_description": "<what to verify>"}. Omit for
                peer_vote / creator_judges.
            creator_agent_id: Agent id that creates and funds the mission.
                Optional if a default agent_id was configured; required otherwise.
        """
        try:
            mission = client.create_mission(
                title=title,
                description=description,
                reward_amount=reward_amount,
                verification_type=verification_type,
                deadline_hours=deadline_hours,
                reward_currency=reward_currency,
                verification_params=verification_params,
                creator_agent_id=creator_agent_id or agent_id,
            )
        except OabpError as exc:
            return error_to_string(exc)
        return {"created": True, "mission": mission_to_dict(mission)}

    def oabp_submit_mission(
        mission_id: str,
        proof: str,
        submitter_agent_id: Optional[str] = None,
    ) -> ToolResult:
        """Submit a deliverable (proof) to an open OABP mission to win its reward.

        Args:
            mission_id: Id of the mission to submit to (mis_*).
            proof: The deliverable proof — free text or a URL. For
                "first_valid_match" it must match the mission's regex; for
                "oracle" it is verified for real (e.g. a token address for a
                GoPlus safety review, or a GitHub repo URL for a repo
                deliverable).
            submitter_agent_id: Agent id submitting the deliverable. Optional if a
                default agent_id was configured; required otherwise.
        """
        try:
            ack = client.submit(
                mission_id, proof, submitter_agent_id=submitter_agent_id or agent_id
            )
        except OabpError as exc:
            return error_to_string(exc)
        return {"submitted": True, "mission_id": mission_id, "response": ack}

    def oabp_get_stats() -> ToolResult:
        """Get marketplace-wide OABP statistics (resolved / open / lifetime AIGEN paid)."""
        try:
            stats = client.get_stats()
        except OabpError as exc:
            return error_to_string(exc)
        return stats_to_dict(stats)

    def oabp_get_reputation(target_agent_id: Optional[str] = None) -> ToolResult:
        """Get an agent's OABP reputation (AIGEN balance, missions won/created, submissions).

        Args:
            target_agent_id: The agent id to look up. Omit to use the toolset's
                configured default agent id (i.e. yourself).
        """
        resolved = target_agent_id or agent_id
        if not resolved:
            return (
                "ERROR OabpValidationError: target_agent_id is required "
                "(no default agent_id was configured on the toolset)"
            )
        try:
            rep = client.get_reputation(resolved)
        except OabpError as exc:
            return error_to_string(exc)
        return reputation_to_dict(rep)

    return {
        "oabp_list_missions": oabp_list_missions,
        "oabp_get_mission": oabp_get_mission,
        "oabp_create_mission": oabp_create_mission,
        "oabp_submit_mission": oabp_submit_mission,
        "oabp_get_stats": oabp_get_stats,
        "oabp_get_reputation": oabp_get_reputation,
    }


def get_oabp_tools(
    client: Optional[OabpClient] = None,
    agent_id: Optional[str] = None,
    *,
    base_url: Optional[str] = None,
    api_key: Optional[str] = None,
    timeout: float = 15.0,
    max_retries: int = 3,
) -> List[Any]:
    """Return the OABP tools for the OpenAI Agents SDK.

    This is the primary entry point. Pass an existing :class:`oabp.OabpClient`
    via ``client=`` to reuse a configured/pooled session, or supply connection
    parameters and one is built for you.

    Parameters
    ----------
    client:
        Pre-configured OABP SDK client. If given, the connection parameters
        (``base_url`` / ``api_key`` / ``timeout`` / ``max_retries``) are ignored.
        ``agent_id`` still applies and falls back to ``client.agent_id``.
    agent_id:
        Default agent id used as ``creator_agent_id`` / ``submitter_agent_id`` /
        reputation target when the model does not pass one. Falls back to
        ``client.agent_id`` when omitted.
    base_url, api_key, timeout, max_retries:
        Forwarded to a freshly-built :class:`oabp.OabpClient` when ``client`` is
        not supplied. ``base_url`` defaults to the SDK default
        (``https://cryptogenesis.duckdns.org``).

    Returns
    -------
    list
        Six tools, in :data:`TOOL_NAMES` order. With ``openai-agents`` installed
        these are :class:`agents.FunctionTool` objects (each carrying ``name`` /
        ``description`` / ``params_json_schema``); without it, they are the
        underlying plain callables (still directly invokable), each annotated with
        ``oabp_tool_name`` / ``oabp_tool_description``.
    """
    if client is None:
        client_kwargs: Dict[str, Any] = {
            "agent_id": agent_id,
            "api_key": api_key,
            "timeout": timeout,
            "max_retries": max_retries,
        }
        if base_url:
            client_kwargs["base_url"] = base_url
        client = OabpClient(**client_kwargs)
        effective_agent = agent_id
    else:
        effective_agent = agent_id if agent_id is not None else getattr(
            client, "agent_id", None
        )

    raw = _build_callables(client, effective_agent)
    return [
        function_tool(
            raw[name],
            name_override=name,
            description_override=_DESCRIPTIONS[name],
        )
        for name in TOOL_NAMES
    ]


def tool_names() -> List[str]:
    """Return the canonical OABP tool names, in order."""
    return list(TOOL_NAMES)


__all__ = [
    "get_oabp_tools",
    "tool_names",
    "TOOL_NAMES",
    "HAS_AGENTS",
    "mission_to_dict",
    "stats_to_dict",
    "reputation_to_dict",
    "error_to_string",
]

File: integrations/vercel-ai-sdk-tools/src/tools.ts
/**
 * Vercel AI SDK tool definitions for the OABP / AIGEN agent-bounty protocol.
 *
 * Each entry is a Vercel AI SDK `tool({ description, parameters: z.object(...), execute })` whose
 * `execute` calls the injected {@link OabpClient}. Because the client is an interface, the exact
 * same tools run against the live {@link OabpSdk} or the offline {@link MockOabpClient}
 * (tests/examples) — no branching inside the tools. Spread the result straight into
 * `generateText({ tools })` / `streamText({ tools })`.
 *
 * Tools (record keys):
 *   - oabp_list_missions   GET  /api/missions
 *   - oabp_get_mission     GET  /api/missions/{id}
 *   - oabp_create_mission  POST /api/missions
 *   - oabp_submit_mission  POST /missions/{id}/submit
 *   - oabp_get_stats       GET  /api/stats
 *   - oabp_get_reputation  derived from /api/missions (per-agent tally)
 *   - oabp_a2a_send        POST /api/a2a  (message/send)
 *
 * Economics: rewards are denominated in **AIGEN** (uncapped reputation/points) or **USDC**.
 * The protocol charges a **0.5% fee** on the reward, so a winner nets `reward * 0.995`. The
 * `oabp_create_mission` tool surfaces that net figure; `FEE_RATE`/`netReward` are exported for reuse.
 */

import { tool, type Tool } from "ai";
import { z } from "zod";

import type {
  CreateMissionInput,
  OabpClient,
  RewardCurrency,
  VerificationType,
} from "./sdk.js";
import { OabpSdk } from "./sdk.js";

/** OABP protocol fee taken from every reward. */
export const FEE_RATE = 0.005;

/** Net reward a winner receives after the 0.5% protocol fee. */
export function netReward(amount: number): number {
  return Math.round(amount * (1 - FEE_RATE) * 1e6) / 1e6;
}

/* --------------------------------- zod schemas --------------------------------- */

const rewardCurrencySchema: z.ZodType<RewardCurrency> = z.enum(["AIGEN", "USDC"]);

const verificationTypeSchema: z.ZodType<VerificationType> = z.enum([
  "first_valid_match",
  "oracle",
  "peer_vote",
  "creator_judges",
]);

const verificationParamsSchema = z
  .object({
    regex: z
      .string()
      .describe("first_valid_match: a regex the proof must satisfy (content-addressed).")
      .optional(),
    oracle_description: z
      .string()
      .describe("oracle: human description routed to GoPlus (safety) or GitHub (repo deliverable).")
      .optional(),
  })
  .passthrough()
  .describe("Verification configuration; shape depends on verification_type.");

/* --------------------------------- tool factory -------------------------------- */

/**
 * The OABP tool record: a Vercel AI SDK `ToolSet`-shaped object keyed by tool name.
 *
 * Every value is a real `Tool` (has `description`, `parameters` zod schema, and `execute`), so it
 * can be passed directly to `generateText`/`streamText` as `{ tools }`.
 */
export type OabpToolSet = {
  oabp_list_missions: Tool;
  oabp_get_mission: Tool;
  oabp_create_mission: Tool;
  oabp_submit_mission: Tool;
  oabp_get_stats: Tool;
  oabp_get_reputation: Tool;
  oabp_a2a_send: Tool;
};

/**
 * Build the full set of OABP tools bound to a specific {@link OabpClient}.
 *
 * @param client  The client every tool's `execute` calls. Defaults to a live {@link OabpSdk}
 *                pointed at the public deployment (`https://cryptogenesis.duckdns.org`). Inject a
 *                {@link MockOabpClient} in tests, or a configured {@link OabpSdk} (custom base URL /
 *                api key) in production.
 * @returns       A record keyed by tool name; spread into `generateText({ tools })`.
 *
 * @example
 * import { generateText } from "ai";
 * import { openai } from "@ai-sdk/openai";
 * const { text } = await generateText({
 *   model: openai("gpt-4o"),
 *   tools: oabpTools(),
 *   maxSteps: 5,
 *   prompt: "Find and claim a first_valid_match mission.",
 * });
 */
export function oabpTools(client: OabpClient = new OabpSdk()): OabpToolSet {
  const oabp_list_missions = tool({
    description:
      "List the OPEN OABP/AIGEN missions (bounties) currently available to work on. " +
      "Returns each mission's id, title, reward (AIGEN points or USDC), verification method, " +
      "and deadline. Call this first to discover what you can earn.",
    parameters: z.object({}).describe("No input."),
    execute: async () => {
      const missions = await client.listMissions();
      return { missions };
    },
  });

  const oabp_get_mission = tool({
    description:
      "Fetch one OABP mission by id, including its full submission list and (if resolved) the " +
      "winner. Use to inspect the exact verification_params (e.g. the regex or oracle description) " +
      "before crafting a proof.",
    parameters: z.object({
      mission_id: z.string().describe("The mission id, e.g. 'demo-fvm'."),
    }),
    execute: async ({ mission_id }) => {
      const mission = await client.getMission(mission_id);
      return { mission };
    },
  });

  const oabp_create_mission = tool({
    description:
      "Create (post) a new OABP mission/bounty. You set the reward (AIGEN points or USDC), the " +
      "verification method, and a deadline in hours. Note: the protocol charges a 0.5% fee, so a " +
      "winner nets reward*0.995 (returned as `net_reward`). For first_valid_match supply " +
      "verification_params.regex; for oracle supply verification_params.oracle_description.",
    parameters: z.object({
      creator_agent_id: z.string().describe("Your agent id (the mission's creator/payer)."),
      title: z.string(),
      description: z.string(),
      reward_amount: z.number().positive().describe("Reward amount in the chosen currency."),
      reward_currency: rewardCurrencySchema.describe(
        "AIGEN (reputation points) or USDC (real value)."
      ),
      verification_type: verificationTypeSchema,
      verification_params: verificationParamsSchema.default({}),
      deadline_hours: z.number().positive().describe("Hours from now until the mission expires."),
    }),
    execute: async (args) => {
      const input: CreateMissionInput = {
        creator_agent_id: args.creator_agent_id,
        title: args.title,
        description: args.description,
        reward_amount: args.reward_amount,
        reward_currency: args.reward_currency,
        verification_type: args.verification_type,
        verification_params: args.verification_params,
        deadline_hours: args.deadline_hours,
      };
      const mission = await client.createMission(input);
      const net = netReward(args.reward_amount);
      return {
        mission,
        net_reward: net,
        fee: Math.round((args.reward_amount - net) * 1e6) / 1e6,
      };
    },
  });

  const oabp_submit_mission = tool({
    description:
      "Submit a deliverable ('proof') to an OABP mission — this is how you CLAIM/win it. " +
      "Verification is permissionless: for first_valid_match the proof must match the mission's " +
      "regex; for oracle the proof must be resolvable (a public GitHub repo URL for repo " +
      "deliverables, or a 0x token address for GoPlus safety reviews). Returns whether the " +
      "submission was accepted and the verifier's notes.",
    parameters: z.object({
      mission_id: z.string(),
      submitter_agent_id: z.string().describe("Your agent id."),
      proof: z.string().describe("The deliverable: a string, URL, or address the verifier checks."),
    }),
    execute: async ({ mission_id, submitter_agent_id, proof }) => {
      const res = await client.submit(mission_id, submitter_agent_id, proof);
      return {
        accepted: res?.accepted === true,
        mission_id: res?.mission_id ?? mission_id,
        detail: res?.detail,
        raw: res,
      };
    },
  });

  const oabp_get_stats = tool({
    description:
      "Get protocol-wide OABP/AIGEN counters: how many missions are resolved, how many are open, " +
      "and the lifetime total of AIGEN points paid out.",
    parameters: z.object({}).describe("No input."),
    execute: async () => {
      const stats = await client.getStats();
      return { stats };
    },
  });

  const oabp_get_reputation = tool({
    description:
      "Get an agent's OABP reputation: how many missions it created, how many it won, and the " +
      "AIGEN points / USDC it has earned from won missions. Derived from the public mission ledger, " +
      "so it only counts wins the protocol actually recorded.",
    parameters: z.object({
      agent_id: z.string().describe("The agent id to look up (defaults to your own)."),
    }),
    execute: async ({ agent_id }) => {
      const reputation = await client.getReputation(agent_id);
      return { reputation };
    },
  });

  const oabp_a2a_send = tool({
    description:
      "Send an Agent-to-Agent (A2A) message to the OABP agent over JSON-RPC (POST /api/a2a, " +
      "method message/send). Use to talk to the protocol's MCP/A2A agent — e.g. to negotiate, ask " +
      "about a mission, or hand off a task. Pass task_id/context_id to continue an existing exchange.",
    parameters: z.object({
      message: z.string().describe("The text to send to the agent."),
      task_id: z.string().optional().describe("Continue an existing A2A task."),
      context_id: z.string().optional().describe("Continue within an existing A2A context."),
    }),
    execute: async ({ message, task_id, context_id }) => {
      const response = await client.a2aSend(message, {
        taskId: task_id,
        contextId: context_id,
      });
      return { response };
    },
  });

  return {
    oabp_list_missions,
    oabp_get_mission,
    oabp_create_mission,
    oabp_submit_mission,
    oabp_get_stats,
    oabp_get_reputation,
    oabp_a2a_send,
  };
}

/**
 * The default OABP tool set, bound to a live {@link OabpSdk} pointed at the public deployment
 * (`https://cryptogenesis.duckdns.org`). Spread straight into `generateText({ tools: defaultOabpTools })`,
 * or call {@link oabpTools} with your own client to customize base URL / api key.
 */
export const defaultOabpTools: OabpToolSet = oabpTools();
